iree/samples/variables_and_state/main.c

```diff
@@ -0,0 +1,182 @@
+
+iree_status_t RunSample(iree_string_view_t bytecode_module_path,
+                        iree_string_view_t driver_name) {
+  //===-------------------------------------------------------------------===//
+  // Instance configuration (this should be shared across sessions).
+  fprintf(stdout, "Configuring IREE runtime instance and '%s' device\n",
+          driver_name.data);
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(IREE_API_VERSION_LATEST,
+                                           &instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_create(
+      &instance_options, iree_allocator_system(), &instance));
+  // TODO(#5724): move device selection into the compiled modules.
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_try_create_default_device(
+      instance, driver_name, &device));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Session configuration (one per loaded module to hold module state).
+  fprintf(stdout, "Creating IREE runtime session\n");
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_session_create_with_device(
+      instance, &session_options, device,
+      iree_runtime_instance_host_allocator(instance), &session));
+  iree_hal_device_release(device);
+
+  fprintf(stdout, "Loading bytecode module at '%s'\n",
+          bytecode_module_path.data);
+  IREE_RETURN_IF_ERROR(iree_runtime_session_append_bytecode_module_from_file(
+      session, bytecode_module_path.data));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Call functions.
+  fprintf(stdout, "Calling functions\n\n");
+  iree_runtime_call_t call_get_value;
+  iree_runtime_call_t call_set_value;
+  iree_runtime_call_t call_add_to_value;
+  iree_runtime_call_t call_reset_value;
+  // Exported function names match what we used in Python.
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.get_value"), &call_get_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.set_value"), &call_set_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.add_to_value"),
+      &call_add_to_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.reset_value"),
+      &call_reset_value));
+
+  iree_hal_buffer_view_t* ret0 = NULL;
+
```

Comment:
including some pseudo code here at the header or above each block would make this easier to skim - can just use the python:
```
// Call functions to manipulate the counter:
// 1. print(get_value())  // initial value
// 2. set_value(101)
...
```

Right now this all just kind of blends together, especially because the variables are reused.